[iOS] Cocoapods Release Proposal Part 1 - Objective-C API wrapper

xta0 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25541 [iOS] Cocoapods Release Proposal Part 1 - Objective-C API wrapper**

## Summary

iOS Cocoapods Release Proposal

- Part1: Objective-C API wrapper
- Part2: TestApps(Objective-C+Swift) + UnitTests
- Part3: Podspec Files
- Part4: CI Changes

The most common way to release a 3rd party framework in iOS is through cocoapods. You can think of cocoapods as a dependency manager like RubyGem or NPM for Node.js. In Pytorch, we're going to release two cocoapods libraries - a C++ static library(libtorch.a) and an Objective-C API wrapper on top of it.

This PR is the first part regarding the Objective-C API binding. They're mostly identical with [Ivan's android Java API wrapper](https://github.com/pytorch/pytorch/pull/25105), except for Objective-C I added some tensor manipulation methods, such as `permute`,`item`,`to`,`view`,etc. Below is an example of doing the inference in Swift:

```swift
func predict(image:UIImage, completion:(_ results:[InferenceResult])->Void){
        guard var data:[Float32] = image.normalizedBuffer() else{
            completion([])
            return
        }
        let imageTensor = TorchTensor.new(with: .float, size: [1,3,224,224], data: UnsafeMutablePointer(&data))
        guard let inputTensor = imageTensor else {
            completion([])
            return
        }
        let imageIValue = TorchIValue.new(with: inputTensor)
        let outputTensor = self.module?.forward([imageIValue])?.toTensor()

        guard let resultTensor = outputTensor else {
            completion([])
            return
        }
        let topKResuls = self.getTopN(results: resultTensor, count: 5)
        completion(topKResuls);
    }
```
## Naming Convention

Since there is no namespace in Objective-C, the convention is to use three capital letters as a prefix. After discussed with David @dreiss and Dmytro @dzhulgakov, we agreed to use `Torch`as the prefix for Objective-C class name.

## Code Structure

```
├── cpp
│   ├── PytorchExp.podspec
│   └── README.md
└── objc
    ├── PytorchExpObjC.podspec
    ├── README.md
    ├── Tests
    │   ├── TestTorchIValue.m
    │   ├── TestTorchModule.m
    │   ├── TestTorchTensor.m
    │   ├── models
    │   └── scripts
    ├── apis
    │   ├── PytorchObjC.h
    │   ├── TorchIValue+Internal.h
    │   ├── TorchIValue.h
    │   ├── TorchModule.h
    │   ├── TorchTensor+Internal.h
    │   ├── TorchTensor.h
    │   └── framework.modulemap
    ├── src
    │   ├── TorchIValue.mm
    │   ├── TorchModule.mm
    │   └── TorchTensor.mm
    └── testApp
        ├── objc
        └── swift
```

- The `apis` folder contains `.h` files. `TorchTensor.h`,`TorchIValue.h` and `TorchModule.h` will be exposed to users. `PytorchObjC.h` is the umbrella header of the Objective-C framework. The `framework.modulemap` is used by cocoapods to create modules for Swift, which will be covered in Part3. 
- The `src` folder contains all the implementation code in Objective-C++.
- The `Tests` folder contains Unit Tests files
- The `testApp` contains two application demos, demonstrating the use of Objective-C APIs

## Test Plan

This PR alone can't be fully tested as it needs cocoapods to be properly installed and setup on the CI machines, which hasn't been done yet. I've tested it using in own Github repo which is just a mirror of the code here, and it worked fine. See [PytorchExp](https://github.com/xta0/PytorchExp/tree/master/ios/objc) for more details.

Differential Revision: [D17185390](https://our.internmc.facebook.com/intern/diff/D17185390)

xta0 committed:
[iOS] Cocoapods Release Proposal Part 1 - Objective-C API wrapper

Summary:

iOS cocoapods release proposal

- Part1: Objective-C API wrapper
- Part2: TestApps(Objective-C+Swift) + UnitTests
- Part3: Podspec files
- Part4: CI changes

The most common way to release a 3rd party framework in iOS is through cocoapods. You can think of cocoapods as a dependency manager like RubyGem or NPM for node.js In Pytorch, we're going to release two cocoapods libraries - a C++ static library(libtorch.a) and an Objective-C API wrapper on top of it.

This PR is the first part regarding the Objective-C API binding. They're mostly identical with Ivan's android Java API wrapper, except for Objective-C I added some tensor manipulation methods, such as `permute`,`item`,`to`,`view`,etc. Below is an example of doing the inference in Swift:

```swift
func predict(image:UIImage, completion:(_ results:[InferenceResult])->Void){
        guard var data:[Float32] = image.normalizedBuffer() else{
            completion([])
            return
        }
        let imageTensor = TorchTensor.new(with: .float, size: [1,3,224,224], data: UnsafeMutablePointer(&data))
        guard let inputTensor = imageTensor else {
            completion([])
            return
        }
        let imageIValue = TorchIValue.new(with: inputTensor)
        let outputTensor = self.module?.forward([imageIValue])?.toTensor()

        guard let resultTensor = outputTensor else {
            completion([])
            return
        }
        let topKResuls = self.getTopN(results: resultTensor, count: 5)
        completion(topKResuls);
    }
```
naming convention:

Since there is no namespace in Objective-C, the convention is to use three capital letters as a prefix. After discussed with David and Dmytro, we agreed to use `Torch`as the prefix for Objective-C class name.

code structure:

```
├── apis
│   ├── PytorchExpObjC.h
│   ├── TorchIValue+Internal.h
│   ├── TorchIValue.h
│   ├── TorchModule.h
│   ├── TorchTensor+Internal.h
│   ├── TorchTensor.h
│   └── framework.modulemap
└── src
    ├── TorchIValue.mm
    ├── TorchModule.mm
    └── TorchTensor.mm
```

The `apis` contains `.h` files. `TorchTensor.h`,`TorchIValue.h` and `TorchModule.h` will be exposed to users. `PytorchObjC.h` is the umbrellar header of the Objective-C framework. The `framework.modulemap` is used by cocoapods to create module for Swift, which will be covered in Part3.

Test Plan:

This PR alone can't be fully tested as it needs cocoapods to be properly installed and setup on the CI machines, which hasn't been done yet. I've tested it using in own Github repo that is just a mirror of the code here, and it worked fine. see PytorchExp for more details.

rename files

yf225 commented on ios/objc/src/TorchTensor.mm:
```diff
@@ -0,0 +1,196 @@
+}
+
+@implementation TorchTensor {
+    std::shared_ptr<at::Tensor> _impl;
```
I think `at::Tensor` / `at::IValue` / `torch::jit::script::Module` are all pimpl types with `intrusive_ptr<T>` as their underlying data structure (e.g. `at::Tensor b = a` would bump the refcount of `a`'s underlying `intrusive_ptr<TensorImpl>`). So we might not need `std::shared_ptr<at::Tensor> _impl` and can just use `at::Tensor _impl`.

dzhulgakov commented on ios/objc/apis/TorchIValue.h:
```diff
@@ -0,0 +1,51 @@
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef NS_ENUM(NSUInteger, TorchIValueType) {
```
would be nice to add string/list of strings too - they might occur for text models quite often

in the future, adding dicts would be nice too, but it's lesser priority

xta0 replied:
I can add string type

dzhulgakov commented on ios/objc/apis/TorchModule.h:
```diff
@@ -0,0 +1,34 @@
+ @param modelPath File path to the torchscript model
+ @return TorchModuel object
+ */
++ (TorchModule* _Nullable)loadTorchscriptModel:(NSString* _Nullable)modelPath;
```
naming: does it need to be a long name like this or is it always prefixed with the class name? if so - why not just `[TorchModule load]`?

xta0 replied:
Unfortunately, +load is a life-cycle method in Objective-C, we don't usually override this method, unless doing some crazy hacking stuff.

dzhulgakov commented on ios/objc/apis/TorchTensor.h:
```diff
@@ -0,0 +1,102 @@
+/**
+ Returns if the tensor has a quntized backend
+ */
+@property(nonatomic,assign, readonly) BOOL quantized;
```
I'd hide it for now as you don't expose other accessors or dtypes for quantized

dzhulgakov commented on ios/objc/apis/TorchTensor.h:
```diff
@@ -0,0 +1,102 @@
+ @param data A raw pointer to a data buffer
+ @return  A tensor object
+ */
++ (nullable TorchTensor* )newWithType:(TorchTensorType)type Size:(NSArray<NSNumber* >*)size Data:(void* _Nullable)data Quantized:(BOOL) quantized;
```
similarly - I'd drop quantized version for now

dzhulgakov commented on ios/objc/apis/TorchTensor.h:
```diff
@@ -0,0 +1,102 @@
+ @param shape The new shape of the tansor
+ @return A new tensor object with the given dimension
+ */
+- (nullable TorchTensor* )permute:(NSArray<NSNumber* >*) shape;
```
why supporting permute, but not other methods? I guess it might be useful but we should be somewhat consistent on which subset we expose

xta0 replied:
These methods will be deleted.

dzhulgakov commented on ios/objc/apis/TorchTensor.h:
```diff
@@ -0,0 +1,102 @@
+
+@end
+
+@interface TorchTensor(ObjectSubscripting)
```
oh, you can override subscripting, nice. Can you override "operator()" too? (I have no clue about ObjC, but it'd be nice to provide on module level)

dzhulgakov commented on ios/objc/apis/TorchTensor.h:
```diff
@@ -0,0 +1,102 @@
+ @param data A raw pointer to a data buffer
+ @return  A tensor object
+ */
++ (nullable TorchTensor* )newWithType:(TorchTensorType)type Size:(NSArray<NSNumber* >*)size Data:(void* _Nullable)data;
```
again, no idea about ObjC - does it support (or does it make sense to provide) overloaded methods that create a Tensor from typed pointer to e.g. floats (and can deduce the types automatically)

dzhulgakov commented on ios/objc/apis/TorchTensor.h:
```diff
@@ -0,0 +1,102 @@
+
+@end
+
+@interface TorchTensor(Operations)
```
I think it'd be very useful to provide a simple way to convert a tensor to a native list of floats/ints/etc. The thing is that subscripting is pretty slow (create a tensor every time). In C++ API we provide TensorAccessor (https://pytorch.org/cppdocs/notes/tensor_basics.html#efficient-access-to-tensor-elements) or one can just do .data(). We should do something similar

dzhulgakov commented on ios/objc/src/TorchTensor.mm:
```diff
@@ -0,0 +1,196 @@
+        dimsVec.push_back(dim);
+    }
+    auto newTensor =  _impl->permute(dimsVec);
+    newTensor.options();
```
what is this?

xta0 replied:
my bad, this is for debug, will be deleted

dzhulgakov commented on ios/objc/src/TorchIValue.mm:
```diff
@@ -0,0 +1,191 @@
+    _(Double, double,double) \
+
+@implementation TorchIValue {
+    std::shared_ptr<at::IValue> _impl;
```
IValue is a tiny type that is intended to be passed by value - there's no need in a shared_ptr (similarly to Tensor)

shoumikhin commented on ios/objc/apis/TorchIValue.h:
```diff
@@ -0,0 +1,51 @@
+
++ (instancetype) newWithNone;
++ (instancetype) newWithTensor:(TorchTensor* )tensor;
++ (instancetype) newWithBool:(NSNumber* )value;
```
```suggestion
+ (instancetype)valueWithBool:(BOOL)value;
```

shoumikhin commented on ios/objc/apis/TorchIValue.h:
```diff
@@ -0,0 +1,51 @@
++ (instancetype) newWithNone;
++ (instancetype) newWithTensor:(TorchTensor* )tensor;
++ (instancetype) newWithBool:(NSNumber* )value;
++ (instancetype) newWithDouble:(NSNumber* )value;
```
```suggestion
+ (instancetype)valueWithDouble:(double)value;
```

shoumikhin commented on ios/objc/apis/TorchIValue.h:
```diff
@@ -0,0 +1,51 @@
++ (instancetype) newWithTensor:(TorchTensor* )tensor;
++ (instancetype) newWithBool:(NSNumber* )value;
++ (instancetype) newWithDouble:(NSNumber* )value;
++ (instancetype) newWithInt:(NSNumber* )value;
```
```suggestion
+ (instancetype)valueWithInt:(int)value;
```

shoumikhin commented on ios/objc/apis/TorchIValue.h:
```diff
@@ -0,0 +1,51 @@
++ (instancetype) newWithBool:(NSNumber* )value;
++ (instancetype) newWithDouble:(NSNumber* )value;
++ (instancetype) newWithInt:(NSNumber* )value;
++ (instancetype) newWithBoolList:(NSArray<NSNumber*>* )value;
```
```suggestion
+ (instancetype)valueWithBoolArray:(NSArray<NSNumber*>*)array;
```

ditto below

shoumikhin commented on ios/objc/apis/TorchTensor+Internal.h:
```diff
@@ -0,0 +1,13 @@
+#import "TorchTensor.h"
```
```objectivec
#import "TorchTensor.h"

#import <Torch/Torch.h>

NS_ASSUME_NONNULL_BEGIN

@interface TorchTensor ()

@property (nonatomic, readonly) at::Tensor nativeTensor;

+ (instancetype)tensorWithNativeTensor:(const at::Tensor&)nativeTensor;

- (instancetype)initWithNativeTensor:(const at::Tensor&)nativeTensor;

@end

NS_ASSUME_NONNULL_END
```

shoumikhin commented on ios/objc/apis/framework.modulemap:
```diff
@@ -0,0 +1,4 @@
+framework module Pytorch {
```
Let's name it `Torch`

xta0 replied:
Unfortunately, `Torch` has been taken

shoumikhin commented on ios/objc/src/TorchIValue.mm:
```diff
@@ -0,0 +1,191 @@
+#import "TorchIValue+Internal.h"
+#import "TorchTensor.h"
+#import "TorchTensor+Internal.h"
+#import <Pytorch/Pytorch.h>
```
The private header for this class should go first, then we want the private header for the tensor, then `std::vector`. Everything else we get for free.

```objectivec
#import "TorchValuePrivate.h"

#import "TorchTensorPrivate.h"

#include <vector>
```

shoumikhin commented on ios/objc/src/TorchIValue.mm:
```diff
@@ -0,0 +1,191 @@
+    _(DoubleList)\
+    _(TensorList)
+
+#define IVALUE_SCALAR_TYPE(_) \
```
The amount of code duplication isn't worth the macros. The code becomes less readable and tedious to debug and modify. Let's just copy-paste a few lines.

shoumikhin commented on ios/objc/src/TorchIValue.mm:
```diff
@@ -0,0 +1,191 @@
++ (instancetype)newWithNone{
+    auto impl = std::make_shared<at::IValue>();
+    if(!impl) {
+        return nil;
```
If we support `nil` return values, the return type must be `nullable instancetype`

shoumikhin commented on ios/objc/src/TorchIValue.mm:
```diff
@@ -0,0 +1,191 @@
+    if(!impl) {
+        return nil;
+    }
+    TorchIValue* value = [TorchIValue new];
```
Prefer `alloc-init` over `new` due to performance implications.

shoumikhin reviewed (commented):
First pass, didn't finish yet.

1) Please run clang-format over the whole codebase.

2) Let's remove the "Py" prefix entirely. Just `Torch`.

3) Private headers should not be exposed.

4) We may want to support Swift Package Manager as well, which requires a particular directory layout:

/Sources/Torch/ *.mm - implementation files and private headers
/Sources/Torch/include/ *.h - public headers (including umbrella)
/Sources/Torch/include/framework.modulemap
/Tests/TorchTests/ *.m - tests for ObjC APIs
/Torch.h - a symlink to the umbrella

yf225 commented on ios/objc/apis/TorchIValue.h:
```diff
@@ -0,0 +1,51 @@
+
+@class TorchTensor;
+
+@interface TorchIValue : NSObject
```
I think here we are trying to mirror the `at::IValue` class name in C++ frontend, and IMO it would be a good idea to have them consistent to avoid confusion.

xta0 commented on ios/objc/apis/TorchIValue+Internal.h:
```diff
@@ -0,0 +1,13 @@
+#import "TorchIValue.h"
+#import <Pytorch/Pytorch.h>
```
`Torch` has been taken in cocoapods

shoumikhin commented on ios/objc/apis/TorchTensor.h:
```diff
@@ -0,0 +1,102 @@
+ @param size The new shape of the tensor
+ @return A new tensor objeect with given size
+ */
+- (nullable TorchTensor* )view:(NSArray<NSNumber* >*)size;
```
Thinking about something like:

```suggestion
- (nullable TorchTensor*)tensorByViewingDimensionsUsingShape:(NSArray<NSNumber* >*)shape;
```

shoumikhin reviewed (commented):
Still looking...

shoumikhin commented on ios/objc/src/TorchModule.mm:
```diff
@@ -0,0 +1,51 @@
+}
+
++ (TorchModule* _Nullable)loadTorchscriptModel:(NSString* _Nullable)modelPath {
+    if(modelPath.length == 0){
```
Let's use asserts for such checks:

```suggestion
    NSParameterAssert(modelPath.length > 0);
```

shoumikhin commented on ios/objc/src/TorchModule.mm:
```diff
@@ -0,0 +1,51 @@
+    auto torchScriptModule = torch::jit::load([modelPath cStringUsingEncoding:NSASCIIStringEncoding]);
+    auto impl = std::make_shared<torch::jit::script::Module>(torchScriptModule);
+    if (!impl) {
+        return nil;
```
Not gonna be `nullptr` here presumably?

shoumikhin commented on ios/objc/src/TorchModule.mm:
```diff
@@ -0,0 +1,51 @@
+}
+
+- (TorchIValue* _Nullable)forward:(NSArray<TorchIValue* >* _Nullable)values {
+    if (values.count == 0){
```
Assert

shoumikhin commented on ios/objc/src/TorchModule.mm:
```diff
@@ -0,0 +1,51 @@
+}
+
+- (TorchIValue* _Nullable)run_method:(NSString* _Nullable)methodName withInputs:(NSArray<TorchIValue* >* _Nullable) values {
+    if (methodName.length == 0 || values.count ==0 ) {
```
Assert

shoumikhin commented on ios/objc/src/TorchModule.mm:
```diff
@@ -0,0 +1,51 @@
+    std::vector<at::IValue> inputs;
+    for(TorchIValue* value in values) {
+        at::IValue atValue = value.toIValue;
+        inputs.push_back(atValue);
```
`emplace_back` for the move semantics?

shoumikhin commented on ios/objc/src/TorchModule.mm:
```diff
@@ -0,0 +1,51 @@
+    for(TorchIValue* value in values) {
+        inputs.push_back(value.toIValue);
+    }
+    if (auto method = _impl->find_method(std::string([methodName cStringUsingEncoding:NSASCIIStringEncoding]))){
```
```suggestion
    if (auto method = _impl->find_method(methodName.UTF8String))){
```

shoumikhin commented on ios/objc/src/TorchTensor.mm:
```diff
@@ -0,0 +1,196 @@
+#import "TorchTensor.h"
+#import "TorchTensor+Internal.h"
+#import <Pytorch/Pytorch.h>
```
The following should be enough:

```suggestion
#import "TorchTensorPrivate.h"
```

shoumikhin commented on ios/objc/src/TorchTensor.mm:
```diff
@@ -0,0 +1,196 @@
+    }
+    std::vector<int64_t> dimsVec;
+    for (auto i = 0; i < size.count; ++i) {
+        int64_t dim = size[i].integerValue;
```
```suggestion
        int64_t dim = size[i].longLongValue;
```

shoumikhin commented on ios/objc/src/TorchTensor.mm:
```diff
@@ -0,0 +1,196 @@
+
+
++ (TorchTensor* )newWithType:(TorchTensorType)type Size:(NSArray<NSNumber* >*)size Data:(void* _Nullable)data Quantized:(BOOL) quantized {
+    if (!data || size.count == 0){
```
Assert

dzhulgakov replied:
size.count == 0 is valid, it's a scalar

also data can be nil if tensor is empty, i.e. one of the dimensions is zero

in general you can skip error checking here and rely on from_blob to do the checks

shoumikhin commented on ios/objc/src/TorchTensor.mm:
```diff
@@ -0,0 +1,196 @@
+    NSAssert(NO, @"Tensors are immutable");
+}
+
+- (instancetype)copyWithZone:(NSZone *)zone {
```
Why do we want to conform to NSCopying at all?

shoumikhin commented on ios/objc/src/TorchTensor.mm:
```diff
@@ -0,0 +1,196 @@
+
++ (TorchTensor* )newWithTensor:(const at::Tensor& ) tensor{
+    std::shared_ptr<at::Tensor> impl = std::make_shared<at::Tensor>(tensor);
+    if(!impl) {
```
Never `nullptr`?

shoumikhin commented on ios/objc/src/TorchTensor.mm:
```diff
@@ -0,0 +1,196 @@
+    for (int i=0; i<dims.size(); ++i){
+        [shapes addObject:@(dims[i])];
+    }
+    t->_size = [shapes copy];
```
No need to copy

shoumikhin commented on ios/objc/src/TorchTensor.mm:
```diff
@@ -0,0 +1,196 @@
+    std::vector<int64_t> dimsVec;
+    for (auto i = 0; i < dims.count; ++i) {
+        int64_t dim = dims[i].integerValue;
+        dimsVec.push_back(dim);
```
Please adjust the surrounding code to let the following replacement compile:

```objectivec
int64_t dimension = shape[index].longLongValue;
nativeShape.emplace_back(dimension);
```

shoumikhin commented on ios/objc/src/TorchTensor.mm:
```diff
@@ -0,0 +1,196 @@
+    }
+    auto newTensor = _impl->view(views);
+    return [TorchTensor newWithTensor:newTensor];
+}
```
This method should look approximately like the following:

```objectivec
- (TorchTensor *)tensorByViewingDimensionsUsingShape:(NSArray<NSNumber* >*)shape {
  NSParameterAssert(shape.count > 0);

  std::vector<int64_t> nativeShape(shape.count);
  for(NSNumber *dimension in shape){
      nativeShape.emplace_back(dimension.longLongValue);
  }
  auto nativeTensor = _impl->view(nativeShape);
  return [[TorchTensor alloc] initWithNativeTensor:nativeTensor];
}
```

shoumikhin reviewed (changes requested):
Finished the first pass

xta0 commented on ios/objc/apis/TorchTensor.h:
```diff
@@ -0,0 +1,102 @@
+
+ @return A NSNumber of object containing the
+ */
+- (nullable NSNumber* )item;
```
This "permute" API will be removed

xta0 commented on ios/objc/apis/TorchTensor.h:
```diff
@@ -0,0 +1,102 @@
+ @param shape The new shape of the tansor
+ @return A new tensor object with the given dimension
+ */
+- (nullable TorchTensor* )permute:(NSArray<NSNumber* >*) shape;
```
This "permute" API will be removed

dzhulgakov commented on ios/objc/apis/TorchIValue.h:
```diff
@@ -0,0 +1,51 @@
+
+@interface TorchIValue : NSObject
+
+@property(nonatomic,assign, readonly) TorchIValueType type;
```
remove this and only provide: `isBool`, `isInt`, `isFloat`, etc

dzhulgakov commented on ios/objc/apis/TorchTensor.h:
```diff
@@ -0,0 +1,102 @@
+/**
+ Data type of the tensor
+ */
+@property(nonatomic,assign, readonly) TorchTensorType type;
```
call it `.dtype`

shoumikhin replied:
`.dataType` then?
Because such shortenings are against Apple's API guidelines for ObjC.

dzhulgakov commented on ios/objc/src/TorchTensor.mm:
```diff
@@ -0,0 +1,196 @@
+}
+
+
++ (TorchTensor* )newWithType:(TorchTensorType)type Size:(NSArray<NSNumber* >*)size Data:(void* _Nullable)data Quantized:(BOOL) quantized {
```
order of arguments - make the same as in C++

dzhulgakov commented on ios/objc/src/TorchTensor.mm:
```diff
@@ -0,0 +1,196 @@
+        int64_t dim = size[i].integerValue;
+        dimsVec.push_back(dim);
+    }
+    at::Tensor tensor = torch::from_blob( (void* )data, dimsVec,scalarTypeFromTensorType(type));
```
here and everywhere: exception safety. We probably should catch exception (all of Torch throws c10::Exception) and wrap it into what ObjC exception mechanism is

shoumikhin replied:
May simply log the error and return `nil`, since that's a constructor method, and returning `nil` is a typical wait to convey failed construction.
Exceptions are sorta deprecated in ObjC for 3rd party devs. Usually there's an out arg of type `NSError` to indicate the method can fail, and sometimes a `BOOL` return value in addition, to indicate success or failure.

xta0 committed:
Update on "[iOS] Cocoapods Release Proposal Part 1 - Objective-C API wrapper"


## Summary

iOS Cocoapods Release Proposal

- Part1: Objective-C API wrapper
- Part2: TestApps(Objective-C+Swift) + UnitTests
- Part3: Podspec Files
- Part4: CI Changes

The most common way to release a 3rd party framework in iOS is through cocoapods. You can think of cocoapods as a dependency manager like RubyGem or NPM for Node.js. In Pytorch, we're going to release two cocoapods libraries - a C++ static library(libtorch.a) and an Objective-C API wrapper on top of it.

This PR is the first part regarding the Objective-C API binding. They're mostly identical with [Ivan's android Java API wrapper](https://github.com/pytorch/pytorch/pull/25105), except for Objective-C I added some tensor manipulation methods, such as `permute`,`item`,`to`,`view`,etc. Below is an example of doing the inference in Swift:

```swift
func predict(image:UIImage, completion:(_ results:[InferenceResult])->Void){
        guard var data:[Float32] = image.normalizedBuffer() else{
            completion([])
            return
        }
        let imageTensor = TorchTensor.new(with: .float, size: [1,3,224,224], data: UnsafeMutablePointer(&data))
        guard let inputTensor = imageTensor else {
            completion([])
            return
        }
        let imageIValue = TorchIValue.new(with: inputTensor)
        let outputTensor = self.module?.forward([imageIValue])?.toTensor()

        guard let resultTensor = outputTensor else {
            completion([])
            return
        }
        let topKResuls = self.getTopN(results: resultTensor, count: 5)
        completion(topKResuls);
    }
```
## Naming Convention

Since there is no namespace in Objective-C, the convention is to use three capital letters as a prefix. After discussed with David @dreiss and Dmytro @dzhulgakov, we agreed to use `Torch`as the prefix for Objective-C class name.

## Code Structure

```
├── cpp
│   ├── PytorchExp.podspec
│   └── README.md
└── objc
    ├── PytorchExpObjC.podspec
    ├── README.md
    ├── Tests
    │   ├── TestTorchIValue.m
    │   ├── TestTorchModule.m
    │   ├── TestTorchTensor.m
    │   ├── models
    │   └── scripts
    ├── apis
    │   ├── PytorchObjC.h
    │   ├── TorchIValue+Internal.h
    │   ├── TorchIValue.h
    │   ├── TorchModule.h
    │   ├── TorchTensor+Internal.h
    │   ├── TorchTensor.h
    │   └── framework.modulemap
    ├── src
    │   ├── TorchIValue.mm
    │   ├── TorchModule.mm
    │   └── TorchTensor.mm
    └── testApp
        ├── objc
        └── swift
```

- The `apis` folder contains `.h` files. `TorchTensor.h`,`TorchIValue.h` and `TorchModule.h` will be exposed to users. `PytorchObjC.h` is the umbrella header of the Objective-C framework. The `framework.modulemap` is used by cocoapods to create modules for Swift, which will be covered in Part3. 
- The `src` folder contains all the implementation code in Objective-C++.
- The `Tests` folder contains Unit Tests files
- The `testApp` contains two application demos, demonstrating the use of Objective-C APIs

## Test Plan

This PR alone can't be fully tested as it needs cocoapods to be properly installed and setup on the CI machines, which hasn't been done yet. I've tested it using in own Github repo which is just a mirror of the code here, and it worked fine. See [PytorchExp](https://github.com/xta0/PytorchExp/tree/master/ios/objc) for more details.

Differential Revision: [D17185390](https://our.internmc.facebook.com/intern/diff/D17185390)

xta0 committed:
Update on "[iOS] Cocoapods Release Proposal Part 1 - Objective-C API wrapper"


## Summary

iOS Cocoapods Release Proposal

- Part1: Objective-C API wrapper
- Part2: TestApps(Objective-C+Swift) + UnitTests
- Part3: Podspec Files
- Part4: CI Changes

The most common way to release a 3rd party framework in iOS is through cocoapods. You can think of cocoapods as a dependency manager like RubyGem or NPM for Node.js. In Pytorch, we're going to release two cocoapods libraries - a C++ static library(libtorch.a) and an Objective-C API wrapper on top of it.

This PR is the first part regarding the Objective-C API binding. They're mostly identical with [Ivan's android Java API wrapper](https://github.com/pytorch/pytorch/pull/25105), except for Objective-C I added some tensor manipulation methods, such as `permute`,`item`,`to`,`view`,etc. Below is an example of doing the inference in Swift:

```swift
func predict(image:UIImage, completion:(_ results:[InferenceResult])->Void){
        guard var data:[Float32] = image.normalizedBuffer() else{
            completion([])
            return
        }
        let imageTensor = TorchTensor.new(with: .float, size: [1,3,224,224], data: UnsafeMutablePointer(&data))
        guard let inputTensor = imageTensor else {
            completion([])
            return
        }
        let imageIValue = TorchIValue.new(with: inputTensor)
        let outputTensor = self.module?.forward([imageIValue])?.toTensor()

        guard let resultTensor = outputTensor else {
            completion([])
            return
        }
        let topKResuls = self.getTopN(results: resultTensor, count: 5)
        completion(topKResuls);
    }
```
## Naming Convention

Since there is no namespace in Objective-C, the convention is to use three capital letters as a prefix. After discussed with David @dreiss and Dmytro @dzhulgakov, we agreed to use `Torch`as the prefix for Objective-C class name.

## Code Structure

```
├── cpp
│   ├── PytorchExp.podspec
│   └── README.md
└── objc
    ├── PytorchExpObjC.podspec
    ├── README.md
    ├── Tests
    │   ├── TestTorchIValue.m
    │   ├── TestTorchModule.m
    │   ├── TestTorchTensor.m
    │   ├── models
    │   └── scripts
    ├── apis
    │   ├── PytorchObjC.h
    │   ├── TorchIValue+Internal.h
    │   ├── TorchIValue.h
    │   ├── TorchModule.h
    │   ├── TorchTensor+Internal.h
    │   ├── TorchTensor.h
    │   └── framework.modulemap
    ├── src
    │   ├── TorchIValue.mm
    │   ├── TorchModule.mm
    │   └── TorchTensor.mm
    └── testApp
        ├── objc
        └── swift
```

- The `apis` folder contains `.h` files. `TorchTensor.h`,`TorchIValue.h` and `TorchModule.h` will be exposed to users. `PytorchObjC.h` is the umbrella header of the Objective-C framework. The `framework.modulemap` is used by cocoapods to create modules for Swift, which will be covered in Part3. 
- The `src` folder contains all the implementation code in Objective-C++.
- The `Tests` folder contains Unit Tests files
- The `testApp` contains two application demos, demonstrating the use of Objective-C APIs

## Test Plan

This PR alone can't be fully tested as it needs cocoapods to be properly installed and setup on the CI machines, which hasn't been done yet. I've tested it using in own Github repo which is just a mirror of the code here, and it worked fine. See [PytorchExp](https://github.com/xta0/PytorchExp/tree/master/ios/objc) for more details.

Differential Revision: [D17185390](https://our.internmc.facebook.com/intern/diff/D17185390)

xta0 committed:
Update on "[iOS] Cocoapods Release Proposal Part 1 - Objective-C API wrapper"


## Summary

iOS Cocoapods Release Proposal

- Part1: Objective-C API wrapper
- Part2: TestApps(Objective-C+Swift) + UnitTests
- Part3: Podspec Files
- Part4: CI Changes

The most common way to release a 3rd party framework in iOS is through cocoapods. You can think of cocoapods as a dependency manager like RubyGem or NPM for Node.js. In Pytorch, we're going to release two cocoapods libraries - a C++ static library(libtorch.a) and an Objective-C API wrapper on top of it.

This PR is the first part regarding the Objective-C API binding. They're mostly identical with [Ivan's android Java API wrapper](https://github.com/pytorch/pytorch/pull/25105), except for Objective-C I added some tensor manipulation methods, such as `permute`,`item`,`to`,`view`,etc. Below is an example of doing the inference in Swift:

```swift
func predict(image:UIImage, completion:(_ results:[InferenceResult])->Void){
        guard var data:[Float32] = image.normalizedBuffer() else{
            completion([])
            return
        }
        let imageTensor = TorchTensor.new(with: .float, size: [1,3,224,224], data: UnsafeMutablePointer(&data))
        guard let inputTensor = imageTensor else {
            completion([])
            return
        }
        let imageIValue = TorchIValue.new(with: inputTensor)
        let outputTensor = self.module?.forward([imageIValue])?.toTensor()

        guard let resultTensor = outputTensor else {
            completion([])
            return
        }
        let topKResuls = self.getTopN(results: resultTensor, count: 5)
        completion(topKResuls);
    }
```
## Naming Convention

Since there is no namespace in Objective-C, the convention is to use three capital letters as a prefix. After discussed with David @dreiss and Dmytro @dzhulgakov, we agreed to use `Torch`as the prefix for Objective-C class name.

## Code Structure

```
├── cpp
│   ├── PytorchExp.podspec
│   └── README.md
└── objc
    ├── PytorchExpObjC.podspec
    ├── README.md
    ├── Tests
    │   ├── TestTorchIValue.m
    │   ├── TestTorchModule.m
    │   ├── TestTorchTensor.m
    │   ├── models
    │   └── scripts
    ├── apis
    │   ├── PytorchObjC.h
    │   ├── TorchIValue+Internal.h
    │   ├── TorchIValue.h
    │   ├── TorchModule.h
    │   ├── TorchTensor+Internal.h
    │   ├── TorchTensor.h
    │   └── framework.modulemap
    ├── src
    │   ├── TorchIValue.mm
    │   ├── TorchModule.mm
    │   └── TorchTensor.mm
    └── testApp
        ├── objc
        └── swift
```

- The `apis` folder contains `.h` files. `TorchTensor.h`,`TorchIValue.h` and `TorchModule.h` will be exposed to users. `PytorchObjC.h` is the umbrella header of the Objective-C framework. The `framework.modulemap` is used by cocoapods to create modules for Swift, which will be covered in Part3. 
- The `src` folder contains all the implementation code in Objective-C++.
- The `Tests` folder contains Unit Tests files
- The `testApp` contains two application demos, demonstrating the use of Objective-C APIs

## Test Plan

This PR alone can't be fully tested as it needs cocoapods to be properly installed and setup on the CI machines, which hasn't been done yet. I've tested it using in own Github repo which is just a mirror of the code here, and it worked fine. See [PytorchExp](https://github.com/xta0/PytorchExp/tree/master/ios/objc) for more details.

Differential Revision: [D17185390](https://our.internmc.facebook.com/intern/diff/D17185390)

xta0 committed:
Update on "[iOS] Cocoapods Release Proposal Part 1 - Objective-C API wrapper"


## Summary

iOS Cocoapods Release Proposal

- Part1: Objective-C API wrapper
- Part2: TestApps(Objective-C+Swift) + UnitTests
- Part3: Podspec Files
- Part4: CI Changes

The most common way to release a 3rd party framework in iOS is through cocoapods. You can think of cocoapods as a dependency manager like RubyGem or NPM for Node.js. In Pytorch, we're going to release two cocoapods libraries - a C++ static library(libtorch.a) and an Objective-C API wrapper on top of it.

This PR is the first part regarding the Objective-C API binding. They're mostly identical with [Ivan's android Java API wrapper](https://github.com/pytorch/pytorch/pull/25105), except for Objective-C I added some tensor manipulation methods, such as `permute`,`item`,`to`,`view`,etc. Below is an example of doing the inference in Swift:

```swift
func predict(image:UIImage, completion:(_ results:[InferenceResult])->Void){
        guard var data:[Float32] = image.normalizedBuffer() else{
            completion([])
            return
        }
        let imageTensor = TorchTensor.new(with: .float, size: [1,3,224,224], data: UnsafeMutablePointer(&data))
        guard let inputTensor = imageTensor else {
            completion([])
            return
        }
        let imageIValue = TorchIValue.new(with: inputTensor)
        let outputTensor = self.module?.forward([imageIValue])?.toTensor()

        guard let resultTensor = outputTensor else {
            completion([])
            return
        }
        let topKResuls = self.getTopN(results: resultTensor, count: 5)
        completion(topKResuls);
    }
```
## Naming Convention

Since there is no namespace in Objective-C, the convention is to use three capital letters as a prefix. After discussed with David @dreiss and Dmytro @dzhulgakov, we agreed to use `Torch`as the prefix for Objective-C class name.

## Code Structure

```
├── cpp
│   ├── PytorchExp.podspec
│   └── README.md
└── objc
    ├── PytorchExpObjC.podspec
    ├── README.md
    ├── Tests
    │   ├── TestTorchIValue.m
    │   ├── TestTorchModule.m
    │   ├── TestTorchTensor.m
    │   ├── models
    │   └── scripts
    ├── apis
    │   ├── PytorchObjC.h
    │   ├── TorchIValue+Internal.h
    │   ├── TorchIValue.h
    │   ├── TorchModule.h
    │   ├── TorchTensor+Internal.h
    │   ├── TorchTensor.h
    │   └── framework.modulemap
    ├── src
    │   ├── TorchIValue.mm
    │   ├── TorchModule.mm
    │   └── TorchTensor.mm
    └── testApp
        ├── objc
        └── swift
```

- The `apis` folder contains `.h` files. `TorchTensor.h`,`TorchIValue.h` and `TorchModule.h` will be exposed to users. `PytorchObjC.h` is the umbrella header of the Objective-C framework. The `framework.modulemap` is used by cocoapods to create modules for Swift, which will be covered in Part3. 
- The `src` folder contains all the implementation code in Objective-C++.
- The `Tests` folder contains Unit Tests files
- The `testApp` contains two application demos, demonstrating the use of Objective-C APIs

## Test Plan

This PR alone can't be fully tested as it needs cocoapods to be properly installed and setup on the CI machines, which hasn't been done yet. I've tested it using in own Github repo which is just a mirror of the code here, and it worked fine. See [PytorchExp](https://github.com/xta0/PytorchExp/tree/master/ios/objc) for more details.

Differential Revision: [D17185390](https://our.internmc.facebook.com/intern/diff/D17185390)

cereallarceny commented:
@xta0 Is there any update on this wrapper? Would love to see this available sometime soon.  😊
